stdlib/public/core/StringUTF8.swift

```diff
@@ -159,7 +159,9 @@ extension String {
       /// True iff the index is at the end of its view or if the next
       /// byte begins a new UnicodeScalar.
       internal var _isOnUnicodeScalarBoundary : Bool {
-        return UTF8._isValidUTF8(UInt32(truncatingBitPattern: _buffer)) || _isAtEnd
+        let buffer = UInt32(truncatingBitPattern: _buffer)
+        let (codePoint, _) = UTF8._decodeOne(buffer)
+        return codePoint != nil || _isAtEnd
       }
```

Comment:
That would change the behavior slightly w.r.t. the original version, as not all non-continuation code units are valid starting code units (e.g. `0xff`). Though the difference is somewhat inconsequential as `_buffer` should be well-formed, and in general it's not clear from the documentation what the handling of mal-formed sequences should be.